File: 8. Const pointers vs pointer to constant.cpp
// Topic: Constant pointers vs pointer to constant
// Description: Show difference between 'int * const' and 'const int *'
#include <iostream>
using namespace std;

int main()
{
    int a = 5;
    int b = 10;
    const int* p_to_const = &a;  // cannot modify *p_to_const
    // *p_to_const = 6; // error
    p_to_const = &b;  // allowed

    int* const const_ptr = &a;  // cannot change const_ptr
    *const_ptr = 7;             // allowed
    // const_ptr = &b; // error

    cout << "p_to_const points to " << *p_to_const << '\n';
    cout << "const_ptr points to " << *const_ptr << '\n';
    return 0;
}

// Time Complexity: O(1).
// Space Complexity: O(1).

File: 8. Const pointers vs pointer to constant.py
"""
Topic: Const pointer vs pointer to constant (Python conceptual demo)
Description: Python doesn't have const; demonstrate immutability and binding differences
"""
def main():
    a = 5
    b = 10
    p_to_const = a  # ints are immutable; cannot change value in-place
    p_to_const = b  # rebinding allowed
    lst = [1]
    const_ptr_like = lst
    const_ptr_like[0] = 2  # can modify content
    print('p_to_const:', p_to_const)
    print('const_ptr_like:', lst)

if __name__ == '__main__':
    main()

# Time Complexity: O(1).
# Space Complexity: O(1).
